docs/source/clickhouse/schema_design.rst

```diff
@@ -0,0 +1,35 @@
+ClickHouse Schema Design Best Practices
+=======================================
+
+Columns based on dictionary data
+--------------------------------
+
+ClickHouse is a columnar datastore, and at run-time it loads columns on-demand
+based on the columns referenced in the query (both the columns ``SELECT`` ed
+and those part of the ``WHERE`` clause). The ability to store different columns independently
+and not load them for every row for every query is part of the performance advantage that
+ClickHouse provides over a traditional RDBMBS (like PostgreSQL).
+
+Commonly, a data schema contains a flexible key:value pair mapping (canonically: ``tags``) and stores that
+data in a column that contains two ``Nested`` arrays where the first array contains the keys
+of the dictionary and the second array contains the values. This works well when
+your dataset design gives you the ability to filter based on other attributes to a small
+number of rows and the flexibility of completely arbitrary keys and values are a real requirement.
+Often, however, a ClickHouse user is interested in rows that contain a specific tag key or a
+specific ``key=value`` pair and in this case one is often best-served by creating a top-level
+column and duplicating (we often call it "promoting" in code) the data.
+
+This duplication is more performance-efficient on two dimensions:
```

Comment:
I think we can simplify the answer to the first bullet point you provided. How does the following sound?

With array accesses Clickhouse needs to find out the index of the key whose value needs to be looked up. That is an O(N) lookup. The same key maybe located at different indexes for each row. Once the index is known. Looking up its value is O(1) since you perform direct index access. If the query has to lookup M rows, this becomes an O(M * N) in terms of time complexity.

With direct column access, there is no index lookup for a key. So the overall complexity to lookup N rows is O(N).

The second point about adding indexes is valid.